### ISSUE TITLE
Frozen chain with link_error and task protocol v1

### ISSUE THREAD

**Issue by brmzkw** [celery/celery#6737]
# Checklist

- [X] I have verified that the issue exists against the `master` branch of Celery.
- [X] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
- [X] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [X] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [X] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [X] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [X] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [X] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [X] I have verified that the issue exists against the `master` branch of Celery.
- [X] I have included the contents of ``pip freeze`` in the issue.
- [X] I have included all the versions of all the external dependencies required
  to reproduce this bug.

#### Possible Duplicates

On IRC, the_drow mentioned it might be related to #6441.

## Environment & Settings

**Celery version**: 5.1.0b1 (singularity)

# Steps to Reproduce

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
$> pip freeze
amqp==5.0.6
billiard==3.6.4.0
-e git+git@github.com:celery/celery.git@8ebcce1523d79039f23da748f00bec465951de2a#egg=celery
click==7.1.2
click-didyoumean==0.0.3
click-plugins==1.1.1
click-repl==0.1.6
kombu==5.1.0b1
prompt-toolkit==3.0.18
pytz==2021.1
six==1.15.0
vine==5.0.0
wcwidth==0.2.5
```

</p>
</details>

## Minimally Reproducible Test Case

This minimal test case only requires Celery, and doesn't need to have anything else installed or running.

```python
import sys

from celery import Celery
from celery.canvas import Signature, chain
from celery.signals import before_task_publish

app = Celery('tasks', broker='pyamqp://guest@localhost//')
app.conf.task_protocol = 1

t1 = Signature('task1')
t2 = Signature('task2')
t3 = Signature('task3')

c = chain(t1, t2)

c.link_error(t3)
c.freeze()

@before_task_publish.connect
def func(body, *args, **kwargs):
    print('--- Task body ---')
    print(body)

    callbacks = body['callbacks']
    print(f'--- {len(callbacks)} callbacks ---')
    for callback in callbacks:
        print(f'\t{callback}')
    sys.exit()

c.apply_async()
```

# Expected Behavior

`body.callbacks` should be a list of one entry.

This minimal test case:
* configures celery to use task protocol version 1
* creates a chain
* calls `chain.link_error()`
* calls `chain.freeze()`.

The field `callbacks` of the message body sent to rabbitmq should be a list of one element: the chain we want to call.

# Actual Behavior

`body.callbacks` is a list of two entries. The first element is the chain with options, the second without options.

First entry:

```
{'task': 'task2', 'args': (), 'kwargs': {}, 'options': {'task_id': '8b6dec14-4086-4340-b8c6-1134775f8fb4', 'reply_to': '02d66300-1bee-3ebb-8b5c-9c7c97a58ffe'}, 'subtask_type': None, 'immutable': False, 'chord_size': None}
```

Second entry:

```
task2()
```

The problem only happens when both `.link_error()` and `.freeze()` are called, and when task protocol version 1 is used:

* [ERROR] if `.link_error()` and `.freeze()` are called, `callbacks` is a list of two elements (this is the test case)
* [OK] if `.link_error()` is not called, and `.freeze()` is called,  `callbacks` is a list of one element
* [OK] if `.link_error()` is called, and `.freeze()` is not called,  `callbacks` is a list of one element
* [OK] when task protocol is left by default (version 2), when `.link_error()` and `.freeze()` are both called,  then `body[2]['chain']` is a list of one element.

The problem has been introduced by ba2dec7956782 : before this commit, `callbacks` was always a list of one element.

Do you have any idea what causes the problem, and what I can do to prevent the tasks from being duplicated?

Thanks,

**Comment by brmzkw** [comment]
Sooo, I spent many hours on this problem.

The problem comes from [Signature.append_to_list_option()](https://github.com/celery/celery/blob/25f6c139e11edd32f5c36542c737ee7c7de2e9cc/celery/canvas.py#L341).

```python
    def append_to_list_option(self, key, value):
        items = self._with_list_option(key)
        if value not in items:
            items.append(value)
        return value
```

In the case a chain calls `.link_error()`, then `.freeze()`, then `.apply_async()` ; the function `append_to_list_option` is called several times, and one of the time with:

* `key = 'link'`
* `value` is a dict with a key `options`, which is a dict containing the key `link_error`
* `items` is a list of dictionaries. One of the dictionaries is exactly equal to `value`, except it doesn't contain the key `link_error`.

The function adds the option since `value is not in items`, because the key `options['link_error']` differ.

A possible patch is to replace:

```python
        hack_value = {
            k: (v if k != 'options' else {
                subk: subv
                for subk, subv in v.items()
                if subk != 'link_error'
            })
            for k, v in dict(value).items()
        }
        if hack_value not in items:
```

In short, we check if the value is already in the list of options, but we ignore the key `options['link_error']`.


The patch is obviously not what needs to be done to fix the problem, but I can't where else to fix the problem.
